### DISCUSSION TITLE
How to use mathjs on the HarmonyOS ?

### DISCUSSION THREAD

**Discussion by wanglifeng95** [josdejong/mathjs#3120] [Q&A]
Mathjs is very useful and I really like it. I can use “ohpm install mathjs” to install and use correctly on HarmonyOS (API9), but it cannot run. I don't know why this is? An example is as follows：

```
import {derivative} from 'mathjs'

@Entry
@Component
struct Index {
  @State message: string = 'Hello World';

  build() {
    Row() {
      Column() {
        Text(derivative('x^2', 'x').toString())
          .fontSize(50)
          .fontWeight(FontWeight.Bold)
      }
      .width('100%')
    }
    .height('100%')
  }
}
```

**Comment by josdejong** [comment]
Is there some error when you try to run it?

**Reply by wanglifeng95** [reply]
The error while running in the HarmonyOS system（API9） is as follows：
```
hvigor ERROR: ArkTsCompilerError [ArkTSCompilerError]: ArkTS:ERROR failed to execute es2abc
    at ModulePreviewMode.throwArkTsCompilerError (C:\Users\flowi\.hvigor\project_caches\51881010c49be65805a24b786b95a4fb\workspace\node_modules\.pnpm\@ohos+hvigor-arkts-compose@3.0.9\node_modules\@ohos\hvigor-arkts-compose\dist\src\data\compose-error.js:1:400)
    at ModulePreviewMode.<anonymous> (D:\programs\Huawei\Sdk\openharmony\9\ets\build-tools\ets-loader\lib\fast_build\ark_compiler\module\module_mode.js:3:510)
    at ChildProcess.emit (node:events:513:28)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:293:12) {
  sdkErrorStack: undefined
}
> hvigor ERROR: Failed :entry:default@PreviewArkTS... 
> hvigor ERROR: default@PreviewArkTS watch work[0] failed.
> hvigor ERROR: BUILD FAILED in 40 s 225 ms 
```


PreviewerLog:
```
[Compile Result]  the size of programs is expected to be 980, but is 979
[Compile Result] Compile error occurred. Fix it based on the above message.
```

**Reply by josdejong** [reply]
Hm, that isn't helpful.

Do you know what kind of JavaScript engine HarmonyOS uses? I can't find anything about that. Only thing that crosses my mind is that HarmonyOS may not support all modern JavaScript features or something like that. You can for example try using the UMD bundle of mathjs, maybe HarmonyOS doesn't support ESM and is trying to use that version.

**Reply by wanglifeng95** [reply]
I currently don't know how to use the UMD bundle of mathjs on harmonyos Deveco Studio. 
I have tried, and the previewer on the HarmonyOS system's API10 can display normally, but there will be errors on its API9. But currently, the real phone of  HarmonyOS(API10) has not been released yet. I will test it again when the real phone of  API10 is released later. 
Thank you for your help.
